### ISSUE TITLE
Trimming the fat out of the JS output

### ISSUE THREAD

**Issue by bitc** [ghcjs/ghcjs#417]
When using GHCJS with some Haskell programs, there is a lot of JavaScript code in the final `all.js` that is not needed.

The Closure compiler removes some, but not all, of the unneeded code.

I have used the [JSCover](http://tntim96.github.io/JSCover/) tool, to analyze my "[Particles Benchmark](https://bitc.github.io/tojs-benchmark/benchmark.html)" program. Of course I am using the minimized output from the closure compiler with the most aggressive settings. (But I have applied a "beauitifier" to the code to add back newlines, in order to get better results from JSCover)

You can view the report [here](https://bitc.github.io/tojs-benchmark/tmp/jscover-report-20150928/jscoverage.html). But these are the most significant results:

- The size of the `all.min.js` file is **248KB**
- JSCover reports that only **13%** of the functions are executed
- JSCover reports that the amount of code that is actually executed is **26%**

This means that there is **184KB** that is unneeded and doesn't need to be compiled into `all.min.js`. (Although I am guessing that a tiny bit of this may be error handling code that does in fact still need to be in there).

The "Particles Benchmark" that was analyzed here does not make any use of threads, I am guessing that a lot of the unneeded code is scheduling related. The JSCover report linked above shows exactly which parts of `all.min.js` are unneeded (red), but it is difficult to understand what they are since all the function and variable names are obfuscated (due to closure compiler).

For reference, "improved-base" is used, and the only modules that are imported are:

- JavaScript.Web.Canvas
- GHCJS.Foreign.Callback
- Control.Monad.State
- Data.List
- Data.Bits

Note that the reason that closure compiler isn't able to eliminate the unneeded functions is that they appear to be called, but the call is inside an "if" or branch that is never actually entered.

I guess the first question is: Is there even an interest in eliminating this 184KB of unneeded overhead (for programs that don't need it)? It is not that large of a number, but it does have a performance impact, especially on mobile, where JS parsing is often slow.

**Comment by spl** [comment]
> Is there even an interest in eliminating this 184KB of unneeded overhead (for programs that don't need it)?

Absolutely!

> It is not that large of a number [...]

I would disagree and say that **74%** is a very large number.

**Comment by ddssff** [comment]
On Mon, Sep 28, 2015 at 10:41 PM, Sean Leather <notifications@github.com>
wrote:

> Is there even an interest in eliminating this 184KB of unneeded overhead
> (for programs that don't need it)?
>
> Absolutely!
>
> It is not that large of a number [...]
>
> I would disagree and say that *74%* is a very large number.
>

​Especially when multiplied by The Internet.​

**Comment by luite** [comment]
> Note that the reason that closure compiler isn't able to eliminate the unneeded functions is that they appear to be called, but the call is inside an "if" or branch that is never actually entered.

That is also the reason that GHCJS doesn't leave out that code. It keeps track of dependencies of all top-level mutually recursive binding groups (called linkable units in the linker, they contain information similar to what a single object file in GHC split-objects does), and links those that are reachable from `main`.

Can you give an example of a branch that is never entered, where this can be decided statically somehow? I'm working on a new code generator which reduces the general code size, but I haven't refined dependency tracking yet. It could be worth taking a look.

**Comment by flip111** [comment]
I had some suggestions [here](https://github.com/reflex-frp/reflex-platform/issues/31) how to rewrite some functions so that they take up less space. But i forgot to open an issue on this repo for it. I'm not sure how well this rewrite works with closure aggressive settings though. Analyzing code for dead code removal is interesting too.

**Comment by eshansingh** [comment]
Is there any progress on this issue?

**Comment by luite** [comment]
yup, but it's quite a bit of work and not quite done. The `tyr` branch is a major overhaul of the code generator to make (among other things) better optimization and smaller output possible. It also has a built-in minifier and source map support. Closure compiler would still make the code a little smaller, but it wouldn't be the night-and-day difference that it is now.

The `dedupe` branch is an implementation of some of the features planned for `tyr` for the current code generator (while working on `Tyr` I realized that some things could be implemented on `Gen2` in a not too invasive way)

It looks like `dedupe` will be ready for general use within a few weeks, with code size reduction of ~40% for bigger projects. I originally planned some partial releases for `Tyr`, but it will depend on how easy it is to get the performance good enough for production use, and whether there are actual improvements over `Gen2`/`dedupe`.

I think `Tyr` is at least a few months out for production use.

**Comment by cdsmith** [comment]
This is great news, Luite!  I've been watching your dedupe branch with a lot of interest.

FYI, I finally gave in and removed the base.jsexe stuff in CodeWorld, because it was causing too many problems with browser caches.  I'd get a bunch of broken programs every time I changed something in the client library.  So I'm now more concerned about code size, too, in the future.

**Comment by luite** [comment]
A downside of the dedupe branch is that it adds an extra pass to the linker. It basically tries to undo some of the code duplication due to GHC's per-module compilation (often inlining or specializing things the same way in multiple modules).

Incremental linking reduces the work for the linker, since there would just be less data to process. The linker could just stop once it hit a symbol in the base set. Would it be doable to make content-dependent URLS for base.jsexe, where the location is derived from a hash of the contents of the file?

**Comment by cdsmith** [comment]
I tried a few things when I last looked at this, including putting some kind of hash of contents into the base.jsexe URL.  I could not figure it out at the time.  Adding that extra path component just turned runtime errors into 404s in some cases.  I'm sure there's some magic browser-cache incantation that fixes this, but I gave up and decided I can worry about other things with more impact for now.

Understood that this increases build times.  That's okay with me for now. :)

**Comment by micahscopes** [comment]
I'm curious about this, any updates?

**Comment by SpencerPark** [comment]
>   The `tyr` branch is a major overhaul of the code generator to make (among other things) better optimization and smaller output possible. It also has a built-in minifier and source map support. ... I think Tyr is at least a few months out for production use.

@luite I am so happy I ran into this. Is there the possibility of this coming to master or one of the upcoming ghc8 branches anytime soon? Source maps would make a huge difference in being able to profile web apps written in ghcjs. Browser tools are pretty good and source maps would help open them up to ghcjs developers. Thanks!

**Comment by danwdart** [comment]
Did this ever happen?
